nutils/types.py

```diff
@@ -290,6 +290,78 @@ class Singleton(Immutable, metaclass=SingletonMeta):
     __eq__ = object.__eq__
 
 
+class DataClassMeta(abc.ABCMeta):
+
+    def __init__(cls, name, bases, namespace, **kwargs):
+        super().__init__(name, bases, namespace, **kwargs)
+        params = []
+        for base in bases:
+            params.extend(base.__signature__.parameters.values())
+        params.extend(inspect.Parameter(name, inspect.Parameter.POSITIONAL_OR_KEYWORD, default=namespace.get(name, inspect.Parameter.empty), annotation=T)
+            for name, T in namespace.get('__annotations__', {}).items())
+        cls.__signature__ = inspect.Signature(params)
+        cls.__cache = weakref.WeakValueDictionary()
+
+    def __call__(cls, *args, **kwargs):
+        bound = cls.__signature__.bind(*args, **kwargs)
+        bound.apply_defaults()
+        if (self := cls.__cache.get(bound.args)) is None:
+            self = object.__new__(cls)
+            self.__dict__.update(bound.arguments)
+            self.__post_init__()
+            cls.__cache[bound.args] = self
+        return self
+
+
+class DataClass(metaclass=DataClassMeta):
+    '''
+    Base class for dataclass-like types, for which an assignment only
+    constructor is inferred from type annotations. Post processing is possible
+    via an optional __post_init__ method. The class adds equality tests,
```

Comment:
``` :meth:`__post_init__` ```

Reply:
fixed

Reply:
correction: sphinx didn't like the change so I changed it to double backticks instead